### DISCUSSION TITLE
HttpClient reuse in Invoke-WebRequest

### DISCUSSION THREAD

**Discussion by CraigChamberlain** [PowerShell/PowerShell#16755] [Q&A] [answered]
I was looking for a good example of the desired lifecycle of an HttpClient in a PowerShell module and landed on the following
[Source](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs).

My original question and analysis of the Cmdlet are on [StackOverflow](https://stackoverflow.com/questions/70720558/best-use-of-httpclient-in-powershell-binary-module/70724931#70724931)

Is it unsafe to reuse an HttpClient across multiple pipeline inputs? I wondered if a client might be instantiated safely in the beginProcess method or if and HttpClientFactory might in someway be provided by the PowerShell host or even module to allow for a little reuse of sockets and resources.

**Comment by iSazonov** [comment, marked as answer]
We discussed this in https://github.com/PowerShell/PowerShell/issues/12764
If I remember right the "fix" could be in explicit re-using `SocketsHttpHandler`.

**Reply by CraigChamberlain** [reply]
Thanks, this discussion is definitely the same.  It's not supper to the point, more about benchmark than how to code an alternative which might be holding it back?

I might look at the code for pwsh5 and see if I could make a fork.

At least for a custom single purpose cmdlet, I discern it would be safe for me to reuse the client at least for each object in a pipeline.  I still don't know if there could be something at module level. 

I understand why a general purpose cmdlet might be more conservative about optimising to make more reliable.  You'd always be way better writing some c# or even an f# script for performance vs PowerShell.
